File: 2_Graph/Part-1 Introduction/L3 Graph Representation _ Java/Gfg.java
//* https://practice.geeksforgeeks.org/problems/print-adjacency-list-1587115620/1?utm_source=youtube&utm_medium=collab_striver_ytdescription&utm_campaign=print-adjacency-list
//* https://youtu.be/OsNklbh9gYI

class Solution {
    // Function to return the adjacency list for each vertex.
    public ArrayList<ArrayList<Integer>> printGraph(
        int V, ArrayList<ArrayList<Integer>> adj) {
        // Code here
        
        ArrayList<ArrayList<Integer>> ans = new ArrayList<ArrayList<Integer>>();
        
        for(int i = 0; i < V; i++){
            ArrayList<Integer> temp = new ArrayList<Integer>();
            temp.add(i);
            
            for(int x : adj.get(i)){
                temp.add(x);
            }
            
            ans.add(temp);
        }
        
        return ans;
    }
}

File: 1_Tree/1. Binary Trees/Part-1 Traversals/L13 Pre, In, and Postorder in one Traversal/Gfg.cpp
//* https://practice.geeksforgeeks.org/problems/postorder-traversal-iterative/1?utm_source=youtube&utm_medium=collab_striver_ytdescription&utm_campaign=postorder-traversal-iterative
//* https://youtu.be/ySp2epYvgTE

vector<int> oneTraversal(Node *root)
{
    vector<int> pre, in, post;

    stack<pair<Node *, int>> st;
    st.push({root, 1});

    while (!st.empty())
    {
        auto it = st.top();
        st.pop();

        if (it.second == 1)
        {
            pre.push_back(it.first->data);
            it.second++;
            st.push(it);

            if (it.first->left != NULL)
            {
                st.push({it.first->left, 1});
            }
        }
        else if (it.second == 2)
        {
            in.push_back(it.first->data);
            it.second++;
            st.push(it);

            if (it.first->right != NULL)
            {
                st.push({it.first->right, 1});
            }
        }
        else
        { // (it.second == 1)
            post.push_back(it.first->data);
        }
    }

    return post;
}
vector<int> postOrder(Node *node)
{
    // code here

    vector<int> ans;
    if (node == NULL)
    {
        return ans;
    }

    // Using Two Stacks->
    // return using_two_stack(node, ans);

    //* Using One Stack->
    // return using_one_stack(node, ans);

    //* Preorder, Inorder, Postorder in One Traversal
    return oneTraversal(node);
}